Unable to link code containing naked functions on ARM64 macOS with ThinLTO enabled

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

_I have originally reported this issue in `wasmtime` repository and was redirected here: https://github.com/bytecodealliance/wasmtime/issues/11957. I have tried to minimize the reproduction, but I failed to do so in reasonable time. The complexity of `wasmtime` crates is critical here. I would appreciate any hints that would help me minimize this._

Compiling Wasmtime 38.0.3 produces a linker error on ARM64 macOS when building with ThinLTO enabled. The error indicates that `wasmtime_fiber_switch` symbol is undefined during linking.

```
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "<426 object files omitted>" "/var/folders/8c/r_zrp9mj12xd414t08zrks940000gn/T/rustc2wYYXA/{libwasmtime-e843f7bbc21927d3,libzstd_sys-7eabaf414ee67d02,libwasmtime_internal_jit_debug-9043bb9a4e496da8}.rlib" "<sysroot>/lib/rustlib/aarch64-apple-darwin/lib/libcompiler_builtins-*.rlib" "-liconv" "-lSystem" "-lc" "-lm" "-arch" "arm64" "-mmacosx-version-min=11.0.0" "-L" "/Users/mk/Developer/software-mansion/scarb/linker-repro/wasmtime-only-test/target/release/build/wasmtime-15671f59b24bbe0f/out" "-L" "/Users/mk/Developer/software-mansion/scarb/linker-repro/wasmtime-only-test/target/release/build/zstd-sys-1ea01473940232bc/out" "-L" "/Users/mk/Developer/software-mansion/scarb/linker-repro/wasmtime-only-test/target/release/build/wasmtime-internal-jit-debug-2aaf81cce8508d62/out" "-o" "/Users/mk/Developer/software-mansion/scarb/linker-repro/wasmtime-only-test/target/release/deps/wasmtime_only_test-69cba5150eff8962" "-Wl,-dead_strip" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld: warning: object file (/private/var/folders/8c/r_zrp9mj12xd414t08zrks940000gn/T/rustc2wYYXA/libwasmtime-e843f7bbc21927d3.rlib[2](242b1992def1ef0b-helpers.o)) was built for newer 'macOS' version (26.0) than being linked (11.0)
          ld: warning: object file (/private/var/folders/8c/r_zrp9mj12xd414t08zrks940000gn/T/rustc2wYYXA/libwasmtime_internal_jit_debug-9043bb9a4e496da8.rlib[2](db3b6bfb95261072-gdbjit.o)) was built for newer 'macOS' version (26.0) than being linked (11.0)
          Undefined symbols for architecture arm64:
            "wasmtime_internal_fiber::stackswitch::aarch64::wasmtime_fiber_switch::h0b241ee887e10750", referenced from:
                wasmtime_internal_fiber::unix::Suspend::switch::hfd4514b4efef3e13 in wasmtime_only_test-69cba5150eff8962.wasmtime-e843f7bbc21927d3.wasmtime.7e3b8021754df0c4-cgu.05.rcgu.o.rcgu.o
          ld: symbol(s) not found for architecture arm64
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          
```

What's important is that this issue only pops up with ThinLTO enabled. It didn't fail with either LTO disabled, set to FatLTO, or Wasmtime 37.

### Steps to Reproduce

- **Architecture**: ARM64 (Apple Silicon)
- **OS**: macOS 26.0
- **Rust**: 1.90.0
- **Wasmtime**: 38.0.3
- **Build Profile**: `release` with `lto = "thin"`

**Cargo.toml:**
```toml
[package]
name = "linker-repro"
version = "0.1.0"
edition = "2021"

[dependencies]
wasmtime = "38"

[profile.release]
lto = "thin"

```

**src/main.rs:**
```rust
fn main() {
    let engine = wasmtime::Engine::default();
    println!("Engine created: {:?}", engine);
}
```

**Build command:**
```bash
cargo build --release
```

This will fail with the linker error shown above.

I think the wasmtime commit that triggers this regression is: https://github.com/bytecodealliance/wasmtime/commit/3e9eca8b23ace6b2631c557e0e84bb8371c18ed2

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

I have tested this reproduction on all stable+beta+nightly at the time of submitting this issue:

`rustc --version --verbose`:
```
rustc 1.90.0 (1159e78c4 2025-09-14)
binary: rustc
commit-hash: 1159e78c4747b02ef996e55082b704c09b970588
commit-date: 2025-09-14
host: aarch64-apple-darwin
release: 1.90.0
LLVM version: 20.1.8
```

```
rustc 1.92.0-beta.1 (3b4dd9bf1 2025-10-28)
binary: rustc
commit-hash: 3b4dd9bf1410f8da6329baa36ce5e37673cbbd1f
commit-date: 2025-10-28
host: aarch64-apple-darwin
release: 1.92.0-beta.1
LLVM version: 21.1.3
```

```
rustc 1.93.0-nightly (292be5c7c 2025-10-29)
binary: rustc
commit-hash: 292be5c7c05138d753bbd4b30db7a3f1a5c914f7
commit-date: 2025-10-29
host: aarch64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.3
```

--- a/compiler/rustc_codegen_ssa/src/mir/naked_asm.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/naked_asm.rs
@@ -181,7 +181,8 @@ fn prefix_and_suffix<'tcx>(
                 }
             }
             Linkage::Internal => {
-                // write nothing
+                // LTO can fail when internal linkage is used.
+                emit_fatal("naked functions may not have internal linkage")
             }
             Linkage::Common => emit_fatal("Functions may not have common linkage"),
             Linkage::AvailableExternally => {

--- a/compiler/rustc_monomorphize/src/partitioning.rs
+++ b/compiler/rustc_monomorphize/src/partitioning.rs
@@ -580,6 +580,16 @@ fn internalize_symbols<'tcx>(
                 }
             }
 
+            // When LTO inlines the caller of a naked function, it will attempt but fail to make the
+            // naked function symbol visible. To ensure that LTO works correctly, do not default
+            // naked functions to internal linkage and default visibility.
+            if let MonoItem::Fn(instance) = item {
+                let flags = cx.tcx.codegen_instance_attrs(instance.def).flags;
+                if flags.contains(CodegenFnAttrFlags::NAKED) {
+                    continue;
+                }
+            }
+
             // If we got here, we did not find any uses from other CGUs, so
             // it's fine to make this monomorphization internal.
             data.linkage = Linkage::Internal;

--- a/tests/assembly-llvm/naked-functions/aix.rs
+++ b/tests/assembly-llvm/naked-functions/aix.rs
@@ -9,7 +9,7 @@
 //@[aix] needs-llvm-components: powerpc
 
 #![crate_type = "lib"]
-#![feature(no_core, asm_experimental_arch, f128, linkage, fn_align)]
+#![feature(no_core, asm_experimental_arch)]
 #![no_core]
 
 // tests that naked functions work for the `powerpc64-ibm-aix` target.

--- a/tests/assembly-llvm/naked-functions/hidden.rs
+++ b/tests/assembly-llvm/naked-functions/hidden.rs
@@ -1,0 +1,41 @@
+//@ revisions: macos-x86 macos-aarch64 linux-x86
+//@ add-minicore
+//@ assembly-output: emit-asm
+//
+//@[macos-aarch64] compile-flags: --target aarch64-apple-darwin
+//@[macos-aarch64] needs-llvm-components: aarch64
+//
+//@[macos-x86] compile-flags: --target x86_64-apple-darwin
+//@[macos-x86] needs-llvm-components: x86
+//
+//@[linux-x86] compile-flags: --target x86_64-unknown-linux-gnu
+//@[linux-x86] needs-llvm-components: x86
+
+#![crate_type = "lib"]
+#![feature(no_core, asm_experimental_arch)]
+#![no_core]
+
+// Tests that naked functions that are not externally linked (e.g. via `no_mangle`)
+// are marked as `Visibility::Hidden` and emit `.private_extern` or `.hidden`.
+//
+// Without this directive, LTO may fail because the symbol is not visible.
+// See also https://github.com/rust-lang/rust/issues/148307.
+
+extern crate minicore;
+use minicore::*;
+
+// CHECK: .p2align 2
+// macos-x86,macos-aarch64: .private_extern
+// linux-x86: .globl
+// linux-x86: .hidden
+// CHECK: ret
+#[unsafe(naked)]
+extern "C" fn ret() {
+    naked_asm!("ret")
+}
+
+// CHECK-LABEL: entry
+#[no_mangle]
+pub fn entry() {
+    ret()
+}

